modules/partition-replicator/src/integrationTest/java/org/apache/ignite/internal/partition/replicator/ItZoneDataReplicationTest.java

```diff
@@ -404,6 +420,99 @@ void testDataRebalance(boolean truncateRaftLog) throws Exception {
         assertThat(kvView2.getAll(null, data2.keySet()), is(data2));
     }
 
+    @ParameterizedTest
+    @ValueSource(booleans = {true, false})
+    void txFinishCommandGetsReplicated(boolean commit) throws Exception {
+        startCluster(3);
+
+        // Create a zone with a single partition on every node.
+        int zoneId = createZone(TEST_ZONE_NAME, 1, cluster.size());
+
+        int tableId1 = createTable(TEST_ZONE_NAME, TEST_TABLE_NAME1);
+        int tableId2 = createTable(TEST_ZONE_NAME, TEST_TABLE_NAME2);
+
+        var zonePartitionId = new ZonePartitionId(zoneId, 0);
+
+        setupTableIdToZoneIdConverter(zonePartitionId, new TablePartitionId(tableId1, 0), new TablePartitionId(tableId2, 0));
+
+        cluster.forEach(Node::waitForMetadataCompletenessAtNow);
+
+        Node node = cluster.get(0);
+
+        setPrimaryReplica(node, zonePartitionId);
+
+        KeyValueView<Integer, Integer> kvView1 = node.tableManager.table(TEST_TABLE_NAME1).keyValueView(Integer.class, Integer.class);
+        KeyValueView<Integer, Integer> kvView2 = node.tableManager.table(TEST_TABLE_NAME2).keyValueView(Integer.class, Integer.class);
+
+        InternalTransaction transaction = unwrapInternalTransaction(node.transactions().begin());
+        kvView1.put(transaction, 42, 69);
+        kvView2.put(transaction, 142, 169);
+        if (commit) {
+            transaction.commit();
+        } else {
+            transaction.rollback();
+        }
+
+        for (Node currentNode : cluster) {
+            assertTrue(waitForCondition(
+                    () -> !txStatesInPartitionStorage(currentNode.txStatePartitionStorage(zoneId, 0)).isEmpty(),
+                    SECONDS.toMillis(10)
+            ));
+        }
+
+        List<CountExpectation> expectations = new ArrayList<>();
+        for (int i = 0; i < cluster.size(); i++) {
+            Node currentNode = cluster.get(i);
+            expectations.add(new CountExpectation(
```

Comment:
Why do we need this `CountExpectation` class? Can we just add a bunch of `Executables` directly?